[SILGen] Insert an unreachable if a function declaration's parameter is uninhabited

theblixguy opened the pull request:
This PR adds a change to `SILGenProlog` (and `SILGenStmt` as well) to create an `unreachable` at the start of a function if any of the function's parameters are uninhabited.

Resolves [SR-8434](https://bugs.swift.org/browse/SR-8434) and [SR-8435](https://bugs.swift.org/browse/SR-8435)

theblixguy committed:
[sema] Check if a function param type is uninhabited and emit a diagnostic if it is

theblixguy committed:
[test] Add tests

mdiep commented:
It can be useful to declare a function that takes an uninhabited type. (To fulfill a protocol requirement, e.g.)

I envisioned a warning if you supplied a _body_ for such a function—the compiler should let you omit a body altogether (https://bugs.swift.org/browse/SR-8434).

i.e. I think this should be valid code:

```swift
// Even though this returns an `Int`, it doesn't have a `return`. It doesn't need one: it's unreachable.
func f(_: Never) -> Int {}
```

But I think should get a warning if you write any code in the function body:

```swift
func f(_: Never) -> Int {
  return 4 // warning: `f` takes an uninhabited parameter. It's code will never be executed.
}
```

theblixguy commented:
> i.e. I think this should be valid code:
> 
> ```swift
> // Even though this returns an `Int`, it doesn't have a `return`. It doesn't need one: it's unreachable.
> func f(_: Never) -> Int {}
> ```
>
> But I think should get a warning if you write any code in the function body:
> 
> ```swift
> func f(_: Never) -> Int {
>   return 4 // warning: `f` takes an uninhabited parameter. It's code will never be executed.
> }
> ```

Interesting, although I am not too sure about it. It seems a bit strange for a function with a return type to not have a return statement 🤔

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3281,6 +3281,17 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
+    // then emit a diagnostic on the func decl, as the function can never
+    // be called.
+    auto params = FD->getParameters()->getArray();
+    for (auto &&index: indices(params)) {
```
Nitpick: You're not using 'index' anywhere in the body, so you can just iterate directly over params:

`for (auto param : *FD->getParameters())`

theblixguy replied:
Fixed!

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3281,6 +3281,17 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
+    auto params = FD->getParameters()->getArray();
+    for (auto &&index: indices(params)) {
+      auto param = params[index];
+      if (param->hasInterfaceType() && param->getType()->isStructurallyUninhabited()) {
```
getInterfaceType() not getType() since its cheaper

theblixguy replied:
I see! Fixed now.

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3281,6 +3281,17 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
+    for (auto &&index: indices(params)) {
+      auto param = params[index];
+      if (param->hasInterfaceType() && param->getType()->isStructurallyUninhabited()) {
+        TC.diagnose(FD->getLoc(), diag::function_parameter_type_uninhabited, FD->getName(), param->getName());
```
Lines should be < 80 characters

theblixguy replied:
Fixed. Now using the parameter's location instead of the function!

theblixguy committed:
[sema] update indentation, remove unused index param, use getInterfaceType instead of getType

mdiep commented:
> It seems a bit strange for a function with a return type to not have a return statement

It's also strange to write a body for a function that can never be executed. 🙂

But I definitely don't think it should be a warning to declare a function that takes an uninhabited parameter: that can be useful.

theblixguy commented:
@jrose-apple Ah! I was wondering if we would need to tweak the type checker as well because otherwise it would complain about a missing return statement.

jrose-apple commented:
I *think* that error comes from SIL-level diagnostics, but I could be wrong.

theblixguy commented:
Hmm, you're right, there's a `diagnoseMissingReturn` and `diagnoseUnreachable` function in Dataflow Diagnostics which deals with a missing return statement.

It seems maybe we can tweak `SILGenEpilog` and/or `SILGenFunction` to insert an `unreachable` if a function declaration contains an uninhabited parameter type. I'll need to take a deeper look at these classes to understand how everything works. Any suggestions @slavapestov?

theblixguy commented:
Hmm, I gave it a shot but it doesn't seem to emit an `unreachable`:

```
if (paramList)
    for (auto *param : *paramList) {
      if (param->getType()->isStructurallyUninhabited()) {
        SILLocation loc(param);
        B.createUnreachable(loc);
      } else {
        emitter.emitParam(param);
      }
      
      if (selfParam) {
        emitter.emitParam(selfParam);
      }
    }
```

The silgen for `func test(x: Never) -> Int {}` already contains an `unreachable` however:

```
// test(x:)
sil hidden @$S5test14test1xSis5NeverO_tF : $@convention(thin) (Never) -> Int {
// %0                                             // user: %1
bb0(%0 : $Never):
  debug_value %0 : $Never, let, name "x", argno 1 // id: %1
  unreachable                                     // id: %2
} // end sil function '$S5test14test1xSis5NeverO_tF'
```

theblixguy commented:
Nevermind, I was running the `swift` installed on my Mac and not the one built by Ninja 🤦‍♂️ Seems to work!

theblixguy committed:
[silgen] Update prolog code to emit unreachable if a func param is uninhabited

theblixguy commented:
At the moment, I am getting a segfault when invoking the frontend with `-emit-silgen`:

```
// test(x:)
sil hidden @$s5test14test1xSis5NeverO_tF : $@convention(thin) (Never) -> Int {
// %0                                             // user: Stack dump:
0.	Program arguments: ./swift -frontend -emit-silgen /Users/suyashsrijan/Desktop/test1.swift
0  swift                    0x000000010a412b98 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 40
1  swift                    0x000000010a411b68 llvm::sys::RunSignalHandlers() + 248
2  swift                    0x000000010a4131b2 SignalHandler(int) + 258
3  libsystem_platform.dylib 0x00007fff78424b3d _sigtramp + 29
4  libsystem_platform.dylib 0x00007fe17782fa98 _sigtramp + 4282429304
5  swift                    0x00000001073e6299 (anonymous namespace)::SILPrinter::print(swift::SILBasicBlock const*) + 473
6  swift                    0x00000001073e8632 swift::SILFunction::print(swift::SILPrintContext&) const + 6402
7  swift                    0x00000001073eaacb swift::SILModule::print(swift::SILPrintContext&, swift::ModuleDecl*, bool) const + 2363
8  swift                    0x000000010655a2a2 writeSIL(swift::SILModule&, swift::PrimarySpecificPaths const&, swift::CompilerInstance&, swift::CompilerInvocation&) + 194
9  swift                    0x00000001065534bb performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 10155
10 swift                    0x000000010654fc7d swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 3021
11 swift                    0x0000000106505f0e main + 686
12 libdyld.dylib            0x00007fff7823b08d start + 1
Segmentation fault: 11
```

theblixguy committed:
[ast] [sema] remove previous changes

theblixguy committed:
[test] add missing newline

jrose-apple commented:
I'm not a SIL expert, but assuming you're building with assertions but didn't get one, that sounds like the SILPrinter is trying to print a null value for something.

theblixguy commented:
@jrose-apple I think so, seems like the basic block is null when `SILPrinter` is trying to print it. The frontend successfully emits the warning and then crashes after it. It may be because I am emitting the `unreachable` in the wrong place (I am using the `param` location) but perhaps I need the location of the basic block's entry point. I could be wrong, but I'll investigate this further soon.

Azoy commented on lib/SILGen/SILGenProlog.cpp:
```diff
@@ -490,11 +490,19 @@ uint16_t SILGenFunction::emitProlog(ParameterList *paramList,
+        unreachableLoc.markAsPrologue();
+        B.createUnreachable(unreachableLoc);
+      }
+      
```
I think you’re emitting the unreachable too early in the functiok emission. I should’ve clarified the specific function I was mentioning. I think you should look at `void SILGenFunction::emitProlog` rather than the one that returns the arg count. In this function, right after you see the `emitProlog` call, is where I think you can check if any parameters have an uninhabited type and emit unreachable.

theblixguy replied:
Thanks for the suggestions, it no longer segfaults now!

theblixguy committed:
[silgen] move unreachable code emission to void emitProlog()

theblixguy committed:
[silgen] Removes whitespace

theblixguy committed:
[test] Updates warning message

theblixguy commented:
@jrose-apple I have fixed the issue and now it works fine! 🎉

mdiep commented on test/decl/func/functions.swift:
```diff
@@ -179,3 +179,9 @@ func parentheticalInout2(_ fn: (((inout Int)), Int) -> ()) {
+func foo(baz: Never) -> Int { // expected-warning{{will never be executed}}
+  print("I can't be called!")
+  return 0
+}
```
Does that mean that `func foo(baz: Never) -> Int {}` works now? If so, could you add a test for it? 😄

theblixguy replied:
It would already contain an `unreachable`. Here's the raw SIL for the above code:

```
// foo(baz:)
sil hidden @test1.foo(baz: Swift.Never) -> Swift.Int : $@convention(thin) (Never) -> Int {
// %0                                             // user: %1
bb0(%0 : $Never):
  debug_value %0 : $Never, let, name "baz", argno 1 // id: %1
  unreachable                                     // id: %2
} // end sil function 'test1.foo(baz: Swift.Never) -> Swift.Int'
```

theblixguy replied:
But it would complain when you invoke the frontend with `emit-ir` about a missing return statement. Let me give it a shot with my change so see if it still complains!

theblixguy replied:
Just tested and it works! Invoking the front-end with `-emit-ir` no longer complains about a missing return. I have updated the tests as well 👍

Commit:
Merge branch 'master' into fix/SR-8435

theblixguy committed:
[test] added another test case

theblixguy commented:
Can someone run swift-ci?

jrose-apple commented on test/decl/func/functions.swift:
```diff
@@ -179,3 +179,12 @@ func parentheticalInout2(_ fn: (((inout Int)), Int) -> ()) {
+
+//===--- Function declaration with uninhabited parameter type
+                                   
+func foo(baz: Never) -> Int { // expected-warning{{will never be executed}}
```
Hm, I wish we could get this to appear on the *next* line, but that doesn't mean we can't take this PR.

theblixguy replied:
I think I may have to fix it as the tests failed because I didn't put the warning on the return statement?

theblixguy replied:
Fixed.

jrose-apple commented:
I'd still like someone more SIL-savvy than me to review, but here's your test run!

@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

theblixguy commented:
> I'd still like someone more SIL-savvy than me to review, but here's your test run!
> 
> @swift-ci Please test

Thanks! Could you tag people those people (or request a review)?

theblixguy commented:
> I'd still like someone more SIL-savvy than me to review, but here's your test run!

Thanks @jrose-apple! Could you tag people those people (or request a review)?

theblixguy committed:
[test] remove tests for now

Commit:
[test] Adds tests for functions with uninhabited params

jrose-apple commented:
@slavapestov or @jckarter, probably.

theblixguy committed:
[test] add missing newline

jckarter commented:
This looks good to me. One small tweak to the diagnostic might be to add a note to remove the function body to quell the warning.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test OS X

theblixguy commented:
Thanks @jckarter! So I should emit a diagnostic (note) after emitting the unreachable?

Also seems like the test is failing:

```
Script:
--
: 'RUN: at line 1';   /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/swift-linux-x86_64/bin/swift -frontend -target x86_64-unknown-linux-gnu  -module-cache-path '/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/clang-module-cache' -swift-version 4  -emit-silgen -verify-syntax-tree -enable-sil-ownership
--
Exit Code: 1

Command Output (stderr):
--
<unknown>:0: error: this mode requires at least one input file

--
```

I think it's because I am missing a `%s` at the end of the RUN command at the top of the test file.

theblixguy committed:
[ast] [diag] [sil] add note for unreachable function body

theblixguy committed:
[silgen] emit diagnostic (note) about unreachable function body

theblixguy committed:
[test] update test case

I think it was failing due to the missing %s at the end of RUN, so I have fixed that as well.

theblixguy commented:
@jckarter I have fixed the test (hopefully!) and added a diagnostic note.

theblixguy committed:
[test] add missing expected-note

mdiep commented on test/decl/func/functions_uninhabited_param.swift:
```diff
@@ -7,4 +7,4 @@ func foo(baz: Never) -> Int { // expected-note {{the function body will never be
 }
 
-func bar(baz: Never) -> Int {} // ok
+func bar(baz: Never) -> Int {} // expected-note {{the function body will never be executed}}
```
This doesn't seem expected. It seems like it should only be a warning if you've put code in the body of the function.

theblixguy replied:
Hmm, seems like I need to check if the closure's body is empty or not before I emit a diagnostic note. I'll fix it now!

theblixguy committed:
[silgen] emit a diagnostic note only if the body is not empty

theblixguy committed:
[test] update tests

theblixguy committed:
[test] move test file to SILGen/

theblixguy committed:
[ast] [diag] [sil] update diagnostic note

theblixguy committed:
[silgen] check if the body is not null before accessing it

theblixguy committed:
[test] update expected-note in test

theblixguy commented:
Can someone run the CI please? Thank you :) cc @jrose-apple @xedin @slavapestov @jckarter

jckarter commented on lib/SILGen/SILGenProlog.cpp:
```diff
@@ -432,6 +433,9 @@ void SILGenFunction::emitProlog(AnyFunctionRef TheClosure,
         B.createUnreachable(unreachableLoc);
+        
+        auto &ctx = SGM.M.getASTContext();
+        ctx.Diags.diagnose(TheClosure.getBody()->getStartLoc(), diag::unreachable_code_uninhabited_param_note);
```
Is it possible to reference the location of the specific parameter that's unreachable, and include its name in the diagnostic here? It'd be nice to say "'param' is of type 'Never' so this function body can never be executed" or something like that.

theblixguy replied:
Done :)

Commit:
[silgen] [ast] adds param name to diagnostics

jckarter commented:
@swift-ci Please test

jrose-apple commented on lib/SILGen/SILGenProlog.cpp:
```diff
@@ -423,6 +424,26 @@ void SILGenFunction::emitProlog(AnyFunctionRef TheClosure,
+        if (auto body = TheClosure.getBody()) {
+          if (!body->getElements().empty()) {
+            auto &ctx = SGM.M.getASTContext();
+            ctx.Diags.diagnose(body->getStartLoc(), diag::unreachable_code_uninhabited_param_note, param->getName());
```
Unfortunately, this isn't allowed: notes must always be emitted following the actual warning or error they should be attached to. This'll have to go at the point of diagnosing the unreachable code.

theblixguy replied:
Oh, do you suggest I put the note there, or just change the note here to a warning?

jckarter replied:
I agree, I'd prefer adding the note to the eventual warning as well, if it's not too much trouble.

theblixguy replied:
Done @jrose-apple @jckarter :)

Commit:
Merge branch 'master' into fix/SR-8435

Commit:
[silgen] Move diagnostic note creation to SILGenStmt

theblixguy committed:
[ast] [diag] fix indentation

Commit:
[silgen] fix indentation

Commit:
Merge branch 'fix/SR-8435' of https://github.com/theblixguy/swift into fix/SR-8435

theblixguy committed:
[silgen] fix brace indent

Commit:
[silgen] run git-clang-format to fix all indentation issues

Commit:
[silgen] fix brace indentation (again)

jrose-apple commented on lib/SILGen/SILGenStmt.cpp:
```diff
@@ -318,6 +318,14 @@ void StmtEmitter::visitBraceStmt(BraceStmt *S) {
+							break;
+						}
+					}
+				}
```
Looks like tabs snuck in instead of spaces here!

theblixguy replied:
I've fixed it by running `git-clang-format`.

jrose-apple commented:
@jckarter should probably take one more look, but this looks good to me. May be worth commenting about why we can assume the unreachable code is due to the function having an uninhabited parameter, though.

@swift-ci Please test

Azoy reviewed (approved):
I know Jordan doesn't particularly like the word `uninhabited` in diagnostics, but this case seems uncommon so I'm okay with this.

jrose-apple commented:
Heh, it slipped by me in looking at the code. That's true; it would probably be nicer to say "is an enum with no cases" or "contains an enum with no cases" as appropriate.

theblixguy commented:
Hmm, I have noticed both "uninhabited" being used in diagnostics (ex: `missing_never_call`) as well as "enum with no cases" (ex: `type_inferred_to_uninhabited_type`), so I am not sure. "Uninhabited" seems succinct (but its type theory jargon) where as "enum with no cases" seems swifty, but in either case, it's the same thing really. I can amend the diagnostic once again if required!

jrose-apple commented:
Unrelated Foundation failure, I think…

@swift-ci Please test Linux

theblixguy commented:
> Unrelated Foundation failure, I think…

Yeah, all good now!

jrose-apple commented:
I don't expect anything to come up here, but

@swift-ci Please test source compatibility

theblixguy commented:
> I don't expect anything to come up here, but

That took a long time haha, but all good here as well 👌

theblixguy commented:
Is this good to merge now? @jrose-apple @jckarter

jckarter commented:
Yeah, looks good. Thanks @theblixguy!

mdiep commented:
Awesome! 🎉
